tests/unit/test_concurrency.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import collections
+import logging
+
+try:
+    from unittest import mock
+except ImportError:  # pragma: NO PY3 COVER
+    import mock
+
+from google.cloud.ndb import _cache
+from google.cloud.ndb import _eventloop
+from google.cloud.ndb import global_cache as global_cache_module
+from google.cloud.ndb import tasklets
+from google.cloud.ndb import utils
+
+
+log = logging.getLogger(__name__)
+
+
+class Delay(object):
+    """A tasklet wrapper which delays the return of a tasklet.
+
+    Used to orchestrate timing of events in async code to test particular scenarios
+    involving concurrency. Use with `mock.patch` to replace particular tasklets with
+    wrapped versions. When those tasklets are called, they will execute and then the
+    wrapper will hang on to the result until :meth:`Delay.advance()` is called, at which
+    time the tasklet's caller will receive the result.
+
+    Args:
+        wrapped (tasklets.Tasklet): The tasklet to be delayed.
+    """
+
+    def __init__(self, wrapped):
+        self.wrapped = wrapped
+        self.info = "Delay {}".format(self.wrapped.__name__)
+        self._futures = collections.deque()
+
+    @tasklets.tasklet
+    def __call__(self, *args, **kwargs):
+        future = tasklets.Future(self.info)
+        self._futures.append(future)
+
+        result = yield self.wrapped(*args, **kwargs)
+        yield future
+        raise tasklets.Return(result)
+
+    def advance(self):
+        """Allow a call to the wrapper to proceed.
+
+        Calls are advanced in the order in which they were orignally made.
+        """
+        self._futures.popleft().set_result(None)
+
+
+def run_until():
+    """Do all queued work on the event loop.
+
+    This will allow any currently running tasklets to execute up to the point that they
+    hit a call to a tasklet that is delayed by :class:`Delay`. When this call is
+    finished, either all in progress tasklets will have been completed, or a call to
+    :class:`Delay.advance` will be required to move execution forward again.
+    """
+    while _eventloop.run1():
+        pass
+
+
+def test_global_cache_concurrent_writes_692(in_context):
```

Comment:
I don't really like this test. I thought in the course of doing this bug fix I'd try to lay some groundwork for #691, with the goal being a pattern that can be used test various concurrency scenarios. While I did get this to work, I'm not crazy about it because:

- It wound up being very "white box", requiring knowledge of what tasklets would be called by `global_lock_for_write` and `global_lock_for_unwrite`.
- It's brittle. It's almost just dumb luck that I could write a test that both reproduced the problem before the fix and verified the fix. The fix could have wound up changing what these functions do enough that this wouldn't have been possible.
- It's difficult to reason about. See how much verbiage goes into just trying to explain what's going on to someone reading the test. We generally don't want tests to require this level of explication. I also spent *a lot* more time writing the test, than I did implementing the fix. 

I've left the test here for now, so that it can maybe used as a jumping off point for inventing a better pattern. I'm open to suggestions.

Reply:
Rather than fighting time, why not take the approach took in my last commit on https://github.com/googleapis/python-ndb/pull/667?  Just make cache calls to represent scenarios you care about, being careful that you use different cache clients for different logical clients.

Speaking of, I'm not sure `_InProcessGlobalCache` is sufficient here. It seems that different cache clients can stomp on each-other's watches.

Reply:
> Rather than fighting time, why not take the approach took in my last commit on #667? Just make cache calls to represent scenarios you care about, being careful that you use different cache clients for different logical clients.
> 

Because in order to expose the bug, you need a specific timing of things happening in parallel calls to `global_lock_for_write` and `global_unlock_for_write`, and you can't reproduce that by just calling them serially. I think I did manage to find a combination of calls inside a tasklet that reproduced the problem, but my worry is that it wouldn't be evident to anyone reading the code why it reproduced the problem. 

After having done the fully deterministic approach, though, maybe I should have stopped there.

> Speaking of, I'm not sure `_InProcessGlobalCache` is sufficient here. It seems that different cache clients can stomp on each-other's watches.

I managed to get the same exception as in the original issue using the diagnosis I had made looking at a live instance using Redis.

Reply:
> > Rather than fighting time, why not take the approach took in my last commit on #667? Just make cache calls to represent scenarios you care about, being careful that you use different cache clients for different logical clients.
> 
> Because in order to expose the bug, you need a specific timing of things happening in parallel calls to `global_lock_for_write` and `global_unlock_for_write`, and you can't reproduce that by just calling them serially. I think I did manage to find a combination of calls inside a tasklet that reproduced the problem, but my worry is that it wouldn't be evident to anyone reading the code why it reproduced the problem.
> 
> After having done the fully deterministic approach, though, maybe I should have stopped there.
> 
> > Speaking of, I'm not sure `_InProcessGlobalCache` is sufficient here. It seems that different cache clients can stomp on each-other's watches.
> 
> I managed to get the same exception as in the original issue using the diagnosis I had made looking at a live instance using Redis.

True

Reply:
> Because in order to expose the bug, you need a specific timing of things happening in parallel calls to `global_lock_for_write` and `global_unlock_for_write`, and you can't reproduce that by just calling them serially. 

True

> > Speaking of, I'm not sure `_InProcessGlobalCache` is sufficient here. It seems that different cache clients can stomp on each-other's watches.
> 
> I managed to get the same exception as in the original issue using the diagnosis I had made looking at a live instance using Redis.

Sure, it just seems that the watch management in `_InProcessGlobalCache` is broken and likely to cause problems.

Reply:
How is it broken? It's supposed to be a decent reference implementation. At any rate, calling out to Redis or Memcached would move this over into the system test realm. I'm still not, overall, happy with where this test is, so I'll be thinking about how to make this better under the auspices of #691

Reply:
> How is it broken?

The watches don't keep track of clients and all clients share the same `_watch_keys` dict.  So if I watch `foo` and then you watch `foo`, my watch will be overwritten by yours.